### PR TITLE
Compiler and linker cleanups

### PR DESCRIPTION
**Pull request by dcbaker**
This is a mixture of cleanups and refactors pulled into one branch for easy review and merging. Most of this came out of my branch to split the dynamic linker methods into it's own class, with a few patches from a static linker cleanup branch with a few new things thrown with. This *should* be a pure refactor, but it's never that easy :)

### CODE REVIEW

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -1394,6 +1277,22 @@ def get_undefined_link_args(self):
...
+
+    def sanitizer_compile_args(self, value: str) -> List[str]:
+        return []
+
+    def sanitizer_link_args(self, value: str) -> List[str]:
+        return []
+
+    @classmethod
+    def get_buildtype_args(cls, buildtype: str) -> List[str]:
+        return self.BUILDTYPE_ARGS[buildtype]
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__ [F821] undefined name 'self' ([link](https://sider.review/gh/repos/19784232/pulls/5393#issue-4073121))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -1590,7 +1498,7 @@ def get_linker_debug_crt_args(self):
         return ['/MDd']
 
     def get_buildtype_args(self, buildtype):
-        args = msvc_buildtype_args[buildtype]
+        args = super().get_buildtype_args(builtype)
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__ [F821] undefined name 'builtype' ([link](https://sider.review/gh/repos/19784232/pulls/5393#issue-4073122))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/d.py [hunk 3]
```diff
@@ -557,6 +566,17 @@ def get_optimization_args(self, optimization_level):
         return gnu_optimization_args[optimization_level]
 
 class LLVMDCompiler(DCompiler):
+
+
+    BUILDTYPE_ARGS = {
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__ [E303] too many blank lines (2) ([link](https://sider.review/gh/repos/19784232/pulls/5393#issue-4073120))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/d.py [hunk 4]
```diff
@@ -494,6 +493,16 @@ def thread_link_flags(self, env):
         return ['-pthread']
 
 class GnuDCompiler(DCompiler):
+
+    BUILDTYPE_ARGS = d{
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__ [E999] SyntaxError: invalid syntax ([link](https://sider.review/gh/repos/19784232/pulls/5393#issue-4073123))

<sub>posted by Sider</sub>

**Comment by jpakkane** [comment]
⚔️ conflicts. Looks good otherwise, though.

**Comment by jpakkane** [comment]
```
OPTLINK (R) for Win32  Release 8.00.17
Copyright (C) Digital Mars 1989-2013  All rights reserved.
http://www.digitalmars.com/ctg/optlink.html
OPTLINK : Warning 9: Unknown Option : NOIPHOBOS64.LIB
dsimpleapp@exe\app.d.obj Offset 00000H Record Type 0064 
 Error 138: Module or Dictionary corrupt 
Error: linker exited with status 1
ninja: build stopped: subcommand failed.
```

Dunno if that is relevant or not. Maybe yes because it is using a wrong linker?

**Comment by dcbaker** [comment]
that does seem relevant. Let's see if I can actually get dmd to install on windows...

**Comment by dcbaker** [comment]
Okay, I think I got it sorted. A mixture of bad inheritance and some cruft from a change I didn't send as part of this series crept in there. I think it's good now.

**Comment by dcbaker** [comment]
@jpakkane, are you happy with this now, or do you want to wait for 0.52?

**Pull request closed without merging**